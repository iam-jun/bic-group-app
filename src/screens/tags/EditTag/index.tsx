import { ExtendedTheme, useTheme } from '@react-navigation/native';
import React, { FC, useState } from 'react';
import { View, StyleSheet, Keyboard } from 'react-native';
import { useRootNavigation } from '~/hooks/navigation';
import { EditTag as EditTagType, ITag } from '~/interfaces/ITag';
import { spacing } from '~/theme';
import useTagsControllerStore from '../store';
import Store from '~/storeRedux';
import modalActions from '~/storeRedux/modal/actions';
import { useBaseHook } from '~/hooks';
import Header from '~/beinComponents/Header';
import TextInput from '~/baseComponents/Input/TextInput';
import Text from '~/baseComponents/Text';

type EditTagProps = {
    route: {
        params: {
          tag: ITag
        },
    }
}

const EditTag: FC<EditTagProps> = (props) => {
  const { params } = props.route || {};
  const { tag } = params || {};
  const { name, id } = tag;
  const { t } = useBaseHook();
  const { rootNavigation } = useRootNavigation();
  const theme = useTheme();
  const { colors } = theme;
  const styles = createStyle(theme);

  const [nameTagState, setNameTagState] = useState(name || '');

  const loading = useTagsControllerStore((state) => state.loading);
  const actions = useTagsControllerStore((state) => state.actions);

  const handleSave = () => {
    const tagUpdate: EditTagType = {
      id,
      name: nameTagState.trim(),
    };
    actions.editTag(tagUpdate);
    Keyboard.dismiss();
  };

  const onSave = () => {
    Store.store.dispatch(modalActions.showAlert({
      title: t('tags:edit_tag'),
      content: t('tags:do_you_want_to_change_this_tag'),
      cancelBtn: true,
      cancelLabel: t('common:btn_cancel'),
      confirmLabel: t('common:btn_edit'),
      onConfirm: () => handleSave(),
    }));
  };

  const isChanged = nameTagState !== name;

  const isDisabledBtnEditTag = nameTagState.trim().length === 0 || !isChanged;

  const handleBack = () => {
    if (isChanged) {
      Keyboard.dismiss();
      Store.store.dispatch(modalActions.showAlert({
        title: t('discard_alert:title'),
        content: t('discard_alert:content'),
        cancelBtn: true,
        cancelLabel: t('common:btn_discard'),
        confirmLabel: t('common:btn_stay_here'),
        onCancel: () => rootNavigation.goBack(),
      }));
      return;
    }
    rootNavigation.goBack();
  };

  const onChangeText = (text: string) => {
    setNameTagState(text.toLowerCase());
  };

  return (
    <View style={styles.container}>
      <Header
        useI18n
        title="tags:edit_tag"
        buttonProps={{ disabled: isDisabledBtnEditTag, loading, style: styles.btnSave }}
        buttonText="common:btn_save"
        onPressButton={onSave}
        onPressBack={handleBack}
      />
      <View style={styles.content}>
        <TextInput
          testID="edit_tag"
          value={nameTagState}
          placeholder={t('tags:input_tag')}
          onChangeText={onChangeText}
          maxLength={32}
          autoCapitalize="none"
        />
        <Text.BodyS color={colors.neutral40} useI18n>tags:input_tag_require_max_length</Text.BodyS>
      </View>
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: colors.white,
    },
    childrenText: {
      paddingVertical: spacing.padding.small,
      paddingBottom: spacing.padding.base,
      paddingHorizontal: spacing.padding.large,
    },
    btnSave: {
      marginRight: spacing.margin.small,
    },
    content: {
      padding: spacing.padding.large,
    },
  });
};

export default EditTag;
