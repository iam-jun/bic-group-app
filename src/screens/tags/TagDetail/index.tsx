import React, { useEffect } from 'react';
import { View, StyleSheet, FlatList } from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';

import { debounce } from 'lodash';
import Header from '~/beinComponents/Header';
import FilterToolbar from '~/components/FilterToolbar';
import useFilterToolbarStore from '~/components/FilterToolbar/store';
import { IPayloadGetSearchPosts } from '~/interfaces/IHome';
import useCommunitiesStore from '~/store/entities/communities';
import useTagDetailStore from './store';
import ContentItem from '~/components/ContentItem';
import spacing from '~/theme/spacing';
import ViewSpacing from '~/beinComponents/ViewSpacing';
import appConfig from '~/configs/appConfig';
import NoSearchResultsFound from '~/components/NoSearchResultsFound';
import LoadingIndicator from '~/beinComponents/LoadingIndicator';

const TagDetail: React.FC<any> = ({ route }) => {
  const { params } = route || {};
  const { tagData, communityId = '' } = params || {};
  const { name: tagName, id: tagId } = tagData || {};

  const theme = useTheme();
  const styles = createStyle(theme);

  const filterPostType = useFilterToolbarStore((state) => state.postType);
  const filterCreatedBy = useFilterToolbarStore((state) => state.createdBy);
  const filterDate = useFilterToolbarStore((state) => state.datePosted);
  const resetFilter = useFilterToolbarStore((state) => state.reset);

  const tagActions = useTagDetailStore((state) => state.actions);
  const loading = useTagDetailStore((state) => state.loading);
  const hasNextPage = useTagDetailStore((state) => state.hasNextPage);
  const resetTag = useTagDetailStore((state) => state.reset);
  const articles = useTagDetailStore((state) => state.articles);

  const community = useCommunitiesStore((state) => state.data[communityId]);
  const { groupId, name: communityName } = community || {};

  useEffect(() => {
    resetTag();
    getData();
  }, [filterPostType, filterCreatedBy, filterDate?.startDate, filterDate?.endDate, groupId, tagId]);

  useEffect(() => () => {
    resetTag();
    resetFilter();
  }, [tagId]);

  const getData = (isLoadMore = false) => {
    const payload: IPayloadGetSearchPosts = {
      searchText: '',
      actors: filterCreatedBy?.id,
      startDate: filterDate?.startDate,
      endDate: filterDate?.endDate,
      groupId,
      tagName,
      type: filterPostType,
    };
    tagActions.getArticles(payload, isLoadMore);
  };

  const onLoadMore = debounce(() => {
    getData(true);
  }, appConfig.searchTriggerTime);

  const renderItem = ({ item }: any) => (
    <ContentItem id={item?.id} testID="tag_detail.article" />
  );

  const renderFooter = () => {
    if (hasNextPage && !loading) {
      return <LoadingIndicator style={{ margin: spacing.margin.small }} />;
    }
    return <ViewSpacing height={spacing.margin.large} />;
  };

  const renderEmpty = () => {
    if (loading) return <LoadingIndicator style={{ margin: spacing.margin.small }} />;
    return <NoSearchResultsFound />;
  };
  const renderItemSeparator = () => <ViewSpacing height={spacing.margin.large} />;

  return (
    <View style={styles.wrapper}>
      <Header
        removeBorderAndShadow
        title={tagName}
        subTitle={communityName}
      />
      <FilterToolbar groupId={groupId} />
      <FlatList
        data={articles}
        showsVerticalScrollIndicator={false}
        style={styles.listContainer}
        renderItem={renderItem}
        ListFooterComponent={renderFooter}
        ListEmptyComponent={renderEmpty}
        ItemSeparatorComponent={renderItemSeparator}
        onEndReached={onLoadMore}
        onEndReachedThreshold={0.2}
        keyExtractor={(item) => `list_articles_by_tag_${item.id}`}
      />
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    wrapper: {
      flex: 1,
      backgroundColor: colors.gray5,
    },
    listContainer: {
      flex: 1,
    },
  });
};

export default TagDetail;
