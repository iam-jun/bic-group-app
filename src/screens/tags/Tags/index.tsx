import { ExtendedTheme, useTheme } from '@react-navigation/native';
import React, { FC, useCallback, useEffect } from 'react';
import {
  View, StyleSheet, Platform, KeyboardAvoidingView,
} from 'react-native';
import Header from '~/beinComponents/Header';
import { ICommunity } from '~/interfaces/ICommunity';
import useCommunitiesStore, { ICommunitiesState } from '~/store/entities/communities';
import CreateTag from './components/CreateTag';
import ViewSpacing from '~/beinComponents/ViewSpacing';
import { spacing } from '~/theme';
import ListTags from './components/ListTags';
import useGroupsStore, { IGroupsState } from '~/store/entities/groups';
import useTagsControllerStore from '../store';

type TagsProps = {
    route: {
        params: {
            id: string;
            type: 'community' | 'group';
        },
    }
}

const Tags: FC<TagsProps> = (props) => {
  const { params } = props.route || {};
  const { id, type } = params || {};

  const { currentGroupId, groups } = useGroupsStore((state: IGroupsState) => state);
  const { group } = groups[currentGroupId] || {};
  const { name: nameGroup, communityId } = group || {};

  const community = useCommunitiesStore(useCallback((
    state: ICommunitiesState,
  ) => state.data[type === 'community' ? id : communityId] || {} as ICommunity, [id, type]));

  const {
    name: nameCommunity, id: idCommunity,
  } = community;

  const theme = useTheme();
  const styles = createStyle(theme);

  const actions = useTagsControllerStore((state) => state.actions);
  const canCUDTag = useTagsControllerStore(useCallback(
    (state) => state.communityCUDTagPermissions[idCommunity],
    [idCommunity],
  ));

  const titleHeader = type === 'community' ? nameCommunity : nameGroup;

  useEffect(() => {
    actions.getCommunityCUDTagPermission(idCommunity);
  }, [idCommunity]);

  return (
    <KeyboardAvoidingView style={styles.keyboardAvoidingView} behavior={Platform.OS === 'ios' ? 'padding' : 'height'}>
      <View style={styles.container}>
        <Header title={titleHeader} />
        {
          canCUDTag && (
            <>
              <ViewSpacing height={spacing.padding.large} />
              <CreateTag communityId={idCommunity} />
            </>
          )
        }
        <ViewSpacing height={spacing.padding.large} />
        <ListTags communityId={idCommunity} />
      </View>
    </KeyboardAvoidingView>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    keyboardAvoidingView: {
      flex: 1,
    },
    container: {
      flex: 1,
      backgroundColor: colors.gray5,
    },
    childrenText: {
      paddingVertical: spacing.padding.small,
      paddingBottom: spacing.padding.base,
      paddingHorizontal: spacing.padding.large,
    },
  });
};

export default Tags;
