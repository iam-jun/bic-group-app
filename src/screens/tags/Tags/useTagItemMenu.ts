/* eslint-disable @typescript-eslint/no-unused-vars */
import i18next from 'i18next';
import { useDispatch } from 'react-redux';
import modalActions from '~/storeRedux/modal/actions';
import Store from '~/storeRedux';

import { useRootNavigation } from '~/hooks/navigation';
import { Button } from '~/baseComponents';
import { ITag } from '~/interfaces/ITag';

const useTagItemMenu = (
  tag: ITag,
) => {
  const { rootNavigation } = useRootNavigation();
  const dispatch = useDispatch();

  const onPressDelete = () => {
    Store.store.dispatch(modalActions.hideBottomList());
    Store.store.dispatch(modalActions.showAlert({
      title: i18next.t('tags:delete_tag_confirm'),
      content: i18next.t('tags:do_you_want_to_delete_this_tag'),
      cancelBtn: true,
      confirmLabel: i18next.t('common:btn_delete'),
      ConfirmBtnComponent: Button.Danger,
      onConfirm: () => {
        // do something
      },
      confirmBtnProps: { type: 'ghost' },
    }));
  };

  const onPressEdit = () => {
    Store.store.dispatch(modalActions.hideBottomList());
    // rootNavigation.navigate(seriesStack.reorderArticles, { tag });
  };

  const defaultData = [
    {
      id: 1,
      title: i18next.t('common:btn_edit'),
      onPress: onPressEdit,
    },
    {
      id: 2,
      title: i18next.t('common:btn_delete'),
      onPress: onPressDelete,
    },
  ];

  const showMenu = () => {
    dispatch(
      modalActions.showBottomList({ isOpen: true, data: defaultData } as any),
    );
  };

  return {
    showMenu,
  };
};

export default useTagItemMenu;
