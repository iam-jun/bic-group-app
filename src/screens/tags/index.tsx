export { default as Tags } from './Tags';
export { default as TagDetail } from './TagDetail';
export { default as EditTag } from './EditTag';
