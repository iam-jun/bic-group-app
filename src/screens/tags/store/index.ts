import { CreateTag } from '~/interfaces/ITag';
import IBaseState, { IBaseListState, InitStateType } from '~/store/interfaces/IBaseState';
import { createStore, resetStore } from '~/store/utils';
import addTag from './actions/addTag';
import deleteTag from './actions/deleteTag';
import { getCommunityTags } from './actions/getCommunityTags';

export interface ITagsController extends IBaseState {
  communityTags: {
    [idCommunity: string]: IBaseListState<string>;
  };
  loading: boolean;
  actions: {
    getCommunityTags: (idCommunity: string, isRefreshing?: boolean) => void;
    addTag: (idCommunity: string, tag: CreateTag) => void;
    deleteTag: (idCommunity: string, idTag: string) => void;
  };
}

export const initState: InitStateType<ITagsController> = {
  communityTags: {},
  loading: false,
};

const tagsControllerStore = (set, get): ITagsController => ({
  ...initState,
  actions: {
    getCommunityTags: getCommunityTags(set, get),
    addTag: addTag(set, get),
    deleteTag: deleteTag(set, get),
  },
  reset: () => resetStore(initState, set),
});

const useTagsControllerStore = createStore<ITagsController>(tagsControllerStore);

export default useTagsControllerStore;
