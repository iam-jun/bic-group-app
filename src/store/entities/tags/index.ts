import { ITag } from '~/interfaces/ITag';
import IBaseState, { InitStateType } from '~/store/interfaces/IBaseState';
import { createStore, resetStore } from '~/store/utils';
import addTags from './actions/addTags';

export interface ITagsState extends IBaseState {
    tags: {
        [id: string]: ITag
    },
    actions: {
        addTags: (data: ITag | ITag[]) => void;
    }
}

const initState: InitStateType<ITagsState> = {
  tags: {},
};

const tagsStore = (set, get): ITagsState => ({
  ...initState,
  actions: {
    addTags: addTags(set, get),
  },
  reset: () => resetStore(initState, set),
});

const useTagsStore = createStore<ITagsState>(tagsStore);

export default useTagsStore;
