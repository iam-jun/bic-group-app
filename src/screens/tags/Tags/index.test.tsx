import * as React from 'react';
import useTagsStore from '~/store/entities/tags';
import { mockTags } from '~/test/mock_data/tags';
import {
  act, renderHook, renderWithRedux,
} from '~/test/testUtils';
import Tags from '.';
import useTagsControllerStore from '../store';

describe('Tags screen', () => {
  it('render correctly with default state', () => {
    const { result: resultUseTagsStore } = renderHook(() => useTagsStore());

    act(() => {
      resultUseTagsStore.current.actions.addTags(mockTags);
      useTagsControllerStore.setState({
        communityTags: {
          123: {
            ids: ['1', '2', '3', '4', '5', '6', '7'],
            loading: false,
            hasNextPage: false,
            refreshing: false,
          },
        },
      });
    });

    const wrapper = renderWithRedux(<Tags route={{
      params: {
        id: '123',
        type: 'community',
      },
    }}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
