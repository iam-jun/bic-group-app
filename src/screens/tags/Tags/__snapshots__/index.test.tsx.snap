// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags screen render correctly with default state 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#EAEDF2",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "zIndex": 2,
        }
      }
      testID="header"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "#FFFFFF",
              "flexDirection": "row",
              "minHeight": 44,
              "paddingTop": 0,
              "paddingVertical": 4,
            },
            Object {
              "elevation": 2,
              "shadowColor": "rgba(0, 0, 0, 1)",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.12,
              "shadowRadius": 1.5,
            },
            undefined,
          ]
        }
        testID="header.content"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "flex": 1,
              "flexDirection": "row",
              "overflow": "hidden",
              "paddingLeft": 8,
              "paddingRight": 8,
            }
          }
        >
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              Object {
                "bottom": 20,
                "left": 20,
                "right": 20,
                "top": 20,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="header.back.button"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                  Object {
                    "alignItems": "center",
                    "height": 44,
                    "justifyContent": "center",
                    "padding": 12,
                    "width": 44,
                  },
                  Object {
                    "backgroundColor": undefined,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    undefined,
                    undefined,
                    undefined,
                  ]
                }
              >
                <View
                  testID="header.back"
                />
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                  "paddingVertical": 8,
                },
                Object {},
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "height": 0,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": 16,
          "paddingHorizontal": undefined,
          "paddingVertical": undefined,
          "width": undefined,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          Object {
            "borderBottomColor": "#E3E6F2",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
            "paddingBottom": 4,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#181C32",
                "fontFamily": "BeVietnamPro-SemiBold",
                "fontSize": 17,
              },
              Object {
                "color": "#181C32",
              },
              undefined,
            ]
          }
        >
          Tags
        </Text>
      </View>
      <RCTScrollView
        ListEmptyComponent={[Function]}
        ListFooterComponent={[Function]}
        contentContainerStyle={
          Object {
            "paddingHorizontal": 16,
          }
        }
        data={Array []}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        refreshControl={
          <RefreshControlMock
            onRefresh={[Function]}
            refreshing={false}
            tintColor="#A4A9B7"
          />
        }
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={Array []}
        viewabilityConfigCallbackPairs={Array []}
      >
        <RCTRefreshControl />
        <View>
          <View
            onLayout={[Function]}
          />
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
